packages/lib/onedrive-api.js

```diff
@@ -136,20 +136,35 @@ class OneDriveApi {
 		}
 	}
 
-	async uploadChunk(url, handle, options) {
+	async uploadChunk(url, handle, content, options) {
 		options = Object.assign({}, options);
 		if (!options.method) { options.method = 'POST'; }
-		if (!options.headers) { options.headers = {}; }
 
-		if (!options.contentLength) throw new Error(' uploadChunk: contentLength is missing');
+		if (!options.contentLength) throw new Error('uploadChunk: contentLength is missing');
+		if (!options.headers) throw new Error('uploadChunk: header is missing');
+
+		if (content) {
+			options.body = (await new Blob([content]).arrayBuffer()).slice(options.startByte, options.startByte + options.contentLength);
+		} else {
+			const chunk = await shim.fsDriver().readFileChunk(handle, options.contentLength);
+			const Buffer = require('buffer').Buffer;
+			const buffer = Buffer.from(chunk, 'base64');
+			options.body = buffer;
+		}
 
-		const chunk = await shim.fsDriver().readFileChunk(handle, options.contentLength);
-		const Buffer = require('buffer').Buffer;
-		const buffer = Buffer.from(chunk, 'base64');
 		delete options.contentLength;
-		options.body = buffer;
+		delete options.startByte;
+
+		let response = null;
+		if (content && shim.isNode) {
+			// couldn't get it working with shim.fetch on desktop but works without problem with the native fetch.
+			// Always received `ECONNRESET` Error. Tried to disable compress option and increased timeout
+			// but didn't help.
+			response = await fetch(url,options);
+		} else {
+			response = await shim.fetch(url,options);
+		}
```

Comment:
I understand why you've done this, but we can't have special cases like this across the codebase with some code using one fetch and other using a different one.

Also that code would bypass the retry and timeout mechanism as well as break in the CLI app, which doesn't have a native fetch. If native fetch is better in the desktop app, perhaps the shim can be updated to use it everywhere instead (but that would require extensive testing)? It would also still need to work with the CLI app which is going to use node-fetch.

Or perhaps the solution here is to use shim.fetchBlob?

Reply:
Yes, I understand that.  I was afraid of that. It was an act of desperation to be honest :D. But I fixed it now. The reason for the error was that the object of options.body was an ArrayBuffer and not a Buffer.

Reply:
>  The reason for the error was that the object of options.body was an ArrayBuffer and not a Buffer.

That's good to know, thanks for looking into @jonath92.